{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "745e89ed",
   "metadata": {},
   "source": [
    "## Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "86a9086f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import random\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "import torch\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "\n",
    "class ConnectionGene:\n",
    "\n",
    "    def __init__(self, in_node_id, out_node_id, is_enabled):\n",
    "        self.in_node_id = in_node_id\n",
    "        self.out_node_id = out_node_id\n",
    "        self.is_enabled = is_enabled\n",
    "        self.innov_num = self._get_correct_innovation_num()\n",
    "        self.weight = None\n",
    "\n",
    "        self.set_rand_weight()\n",
    "\n",
    "    def set_weight(self, new_weight):\n",
    "        \"\"\"\n",
    "        Sets new weight\n",
    "        :param new_weight: type float\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "        self.weight = torch.Tensor([new_weight]).to(device)\n",
    "\n",
    "    def set_rand_weight(self):\n",
    "        \"\"\"\n",
    "        Weight is set to a random value\n",
    "        :return: None - modifies object\n",
    "        \"\"\"\n",
    "        self.weight = torch.Tensor(torch.normal(torch.arange(0, 1).float())).to(device)\n",
    "\n",
    "    def set_innov_num(self, num):\n",
    "        \"\"\"\n",
    "        Only use when copying a gene to avoid speciation issues\n",
    "        :return: None - modifies object\n",
    "        \"\"\"\n",
    "        self.innov_num = num\n",
    "\n",
    "    def _get_correct_innovation_num(self):\n",
    "        # This method keeps track of a generation's innovations\n",
    "        for connect_gene in Population.current_gen_innovation:\n",
    "            if self == connect_gene:\n",
    "                return connect_gene.innov_num\n",
    "        # Is new innovation\n",
    "        Population.current_gen_innovation.append(self)\n",
    "        return Population.get_new_innovation_num()\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        return (self.in_node_id == other.in_node_id) and (self.out_node_id == other.out_node_id)\n",
    "\n",
    "    def __str__(self):\n",
    "        return 'In: ' + str(self.in_node_id) + '\\nOut: ' + str(self.out_node_id) + '\\nIs Enabled: ' + str(self.is_enabled) + '\\nInnovation #: ' + str(self.innov_num) + '\\nWeight: ' + str(float(self.weight)) + '\\n'\n",
    "\n",
    "class NodeGene:\n",
    "    def __init__(self, node_id, node_type):\n",
    "        self.id = node_id\n",
    "        self.type = node_type\n",
    "        self.unit = None\n",
    "\n",
    "    def __str__(self):\n",
    "        return str(self.id) + '-' + self.type\n",
    "\n",
    "class Genome:\n",
    "\n",
    "    def __init__(self):\n",
    "        self.connection_genes = []\n",
    "        self.node_genes = []\n",
    "        self.node_ids = set()\n",
    "        self.innov_nums = set()\n",
    "        self.fitness = None\n",
    "        self.adjusted_fitness = None\n",
    "        self.species = None\n",
    "\n",
    "    def add_connection_mutation(self):\n",
    "        \"\"\"\n",
    "        In the add connection mutation, a single new connection gene is added\n",
    "        connecting two previously unconnected nodes.\n",
    "        \"\"\"\n",
    "        # TODO: better algorithm?\n",
    "\n",
    "        potential_inputs = [n.id for n in self.node_genes if n.type != 'output']\n",
    "        potential_outputs = [n.id for n in self.node_genes if n.type != 'input' and n.type != 'bias']\n",
    "\n",
    "        if len(potential_outputs) != 0 and len(potential_inputs) != 0:\n",
    "            node_in_id = random.choice(potential_inputs)\n",
    "            node_out_id = random.choice(potential_outputs)\n",
    "\n",
    "            if self._is_valid_connection(node_in_id, node_out_id):\n",
    "                # Add connection\n",
    "                self.add_connection_gene(node_in_id, node_out_id)\n",
    "\n",
    "    def add_node_mutation(self):\n",
    "        \"\"\"\n",
    "        This method adds a node by modifying connection genes.\n",
    "        In the add node mutation an existing connection is split and the new node placed where the old\n",
    "        connection used to be. The old connection is disabled and two new connections are added to the genotype.\n",
    "        The new connection leading into the new node receives a weight of 1, and the new connection leading\n",
    "        out receives the same weight as the old connection.\n",
    "        \"\"\"\n",
    "        # Get new_node id\n",
    "        new_node = self.add_node_gene('hidden')\n",
    "        # Get a random existing connection\n",
    "        existing_connection = self._get_rand_connection_gene()\n",
    "\n",
    "        # Create two connections to replace existing connection with new node in the middle\n",
    "        self.add_connection_gene(existing_connection.in_node_id, new_node.id, weight=1)\n",
    "        self.add_connection_gene(new_node.id, existing_connection.out_node_id, weight=existing_connection.weight)\n",
    "\n",
    "        # disable original connection\n",
    "        existing_connection.is_enabled = False\n",
    "\n",
    "    def get_num_excess_genes(self, other):\n",
    "        num_excess = 0\n",
    "\n",
    "        max_innov_num = max(other.innov_nums)\n",
    "        for c_gene in self.connection_genes:\n",
    "            if c_gene.innov_num > max_innov_num:\n",
    "                num_excess += 1\n",
    "\n",
    "        max_node_id = max([n.id for n in other.node_genes])\n",
    "        for n in self.node_genes:\n",
    "            if n.id > max_node_id:\n",
    "                num_excess += 1\n",
    "\n",
    "        return num_excess\n",
    "\n",
    "    def get_num_disjoint_genes(self, other):\n",
    "        num_disjoint = 0\n",
    "\n",
    "        max_innov_num = max(other.innov_nums)\n",
    "        for c_gene in self.connection_genes:\n",
    "            if c_gene.innov_num <= max_innov_num:\n",
    "                if other.get_connect_gene(c_gene.innov_num) is None:\n",
    "                    num_disjoint += 1\n",
    "\n",
    "        max_node_id = max([n.id for n in other.node_genes])\n",
    "        for n in self.node_genes:\n",
    "            if n.id <= max_node_id:\n",
    "                if other.get_node_gene(n.id) is None:\n",
    "                    num_disjoint += 1\n",
    "\n",
    "        return num_disjoint\n",
    "\n",
    "    def get_connect_gene(self, innov_num):\n",
    "        for c_gene in self.connection_genes:\n",
    "            if c_gene.innov_num == innov_num:\n",
    "                return c_gene\n",
    "        return None\n",
    "\n",
    "    def get_node_gene(self, id):\n",
    "        for n in self.node_genes:\n",
    "            if n.id == id:\n",
    "                return n\n",
    "        return None\n",
    "\n",
    "    def get_avg_weight_difference(self, other):\n",
    "        weight_difference = 0.0\n",
    "        num_weights = 0.0\n",
    "\n",
    "        for c_gene in self.connection_genes:\n",
    "            matching_gene = other.get_connect_gene(c_gene.innov_num)\n",
    "            if matching_gene != None:\n",
    "                weight_difference += float(c_gene.weight) - float(matching_gene.weight)\n",
    "                num_weights += 1\n",
    "\n",
    "        if num_weights == 0.0:\n",
    "            num_weights = 1.0\n",
    "        return weight_difference / num_weights\n",
    "\n",
    "    def get_inputs_ids(self, node_id):\n",
    "        \"\"\"\n",
    "        :param node_id: A node's id\n",
    "        :return: An array of the ids of each node who's output is an input to the node_id param\n",
    "        \"\"\"\n",
    "        node_input_ids = []\n",
    "        for c_gene in self.connection_genes:\n",
    "            if (c_gene.out_node_id == node_id) and c_gene.is_enabled:\n",
    "                node_input_ids.append(c_gene.in_node_id)\n",
    "        return node_input_ids\n",
    "\n",
    "    def add_connection_gene(self, in_node_id, out_node_id, is_enabled=True, weight=None):\n",
    "        new_c_gene = ConnectionGene(in_node_id, out_node_id, is_enabled)\n",
    "\n",
    "        if weight != None:\n",
    "            new_c_gene.set_weight(float(weight))\n",
    "\n",
    "        self.connection_genes.append(new_c_gene)\n",
    "        # Maintain Genome attributes\n",
    "        self.node_ids.add(in_node_id)\n",
    "        self.node_ids.add(out_node_id)\n",
    "        self.innov_nums.add(new_c_gene.innov_num)\n",
    "\n",
    "    def add_node_gene(self, n_type):\n",
    "        new_id = len(self.node_genes)\n",
    "        new_gene = NodeGene(new_id, n_type)\n",
    "        self.node_genes.append(new_gene)\n",
    "        return new_gene\n",
    "\n",
    "    def add_connection_copy(self, copy):\n",
    "        new_c_gene = ConnectionGene(copy.in_node_id, copy.out_node_id, copy.is_enabled)\n",
    "        new_c_gene.set_weight(float(copy.weight))\n",
    "        new_c_gene.set_innov_num(copy.innov_num)\n",
    "\n",
    "        self.connection_genes.append(new_c_gene)\n",
    "        # Maintain Genome attributes\n",
    "        self.node_ids.add(copy.in_node_id)\n",
    "        self.node_ids.add(copy.out_node_id)\n",
    "        self.innov_nums.add(new_c_gene.innov_num)\n",
    "\n",
    "    def add_node_copy(self, copy):\n",
    "        self.node_genes.append(NodeGene(copy.id, copy.type))\n",
    "\n",
    "    def get_connections_in(self, node_id):\n",
    "        \"\"\"\n",
    "        :return: the connection genes in to the node identified by the :param: node_id\n",
    "        \"\"\"\n",
    "        genes = []\n",
    "        for gene in self.connection_genes:\n",
    "            if (gene.out_node_id == node_id) and gene.is_enabled:\n",
    "                genes.append(gene)\n",
    "        return genes\n",
    "\n",
    "    def _get_rand_node_id(self):\n",
    "        return random.choice(list(self.node_ids))\n",
    "\n",
    "    def _get_rand_connection_gene(self):\n",
    "        return random.choice(self.connection_genes)\n",
    "\n",
    "    def _get_connections_out(self, node_id):\n",
    "        \"\"\"\n",
    "        :return: the connection genes out of the node identified by the :param: node_id\n",
    "        \"\"\"\n",
    "        genes = []\n",
    "        for gene in self.connection_genes:\n",
    "            if (gene.in_node_id == node_id) and gene.is_enabled:\n",
    "                genes.append(gene)\n",
    "        return genes\n",
    "\n",
    "    def creates_cycle(self, node_in_id, node_out_id):\n",
    "        \"\"\"\n",
    "        Checks if the addition of a connection gene will create a cycle in the computation graph\n",
    "        :param node_in_id: In node of the connection gene\n",
    "        :param node_out_id: Out node of the connection gene\n",
    "        :return: Boolean value\n",
    "        \"\"\"\n",
    "        if node_in_id == node_out_id:\n",
    "            return True\n",
    "\n",
    "        visited = {node_out_id}\n",
    "        while True:\n",
    "            num_added = 0\n",
    "\n",
    "            for c_gene in self.connection_genes:\n",
    "                if c_gene.in_node_id in visited and c_gene.out_node_id not in visited:\n",
    "\n",
    "                    if c_gene.out_node_id == node_in_id:\n",
    "                        return True\n",
    "                    else:\n",
    "                        visited.add(c_gene.out_node_id)\n",
    "                        num_added += 1\n",
    "\n",
    "            if num_added == 0:\n",
    "                return False\n",
    "\n",
    "    def _is_valid_connection(self, node_in_id, node_out_id):\n",
    "        does_creates_cycle = self.creates_cycle(node_in_id, node_out_id)\n",
    "        does_connection_exist = self._does_connection_exist(node_in_id, node_out_id)\n",
    "\n",
    "        return (not does_creates_cycle) and (not does_connection_exist)\n",
    "\n",
    "    def _does_connection_exist(self, node_1_id, node_2_id):\n",
    "        for c_gene in self.connection_genes:\n",
    "            if (c_gene.in_node_id == node_1_id) and (c_gene.out_node_id == node_2_id):\n",
    "                return True\n",
    "            elif (c_gene.in_node_id == node_2_id) and (c_gene.out_node_id == node_1_id):\n",
    "                return True\n",
    "        return False\n",
    "\n",
    "    def get_outputs(self, node, nodes):\n",
    "        \"\"\"\n",
    "        Gets an unordered list of the node ids n_id outputs to\n",
    "        :param node: The node who's output nodes are being retrieved\n",
    "        :param nodes: List containing genome's node genes\n",
    "        :return: List of node genes\n",
    "        \"\"\"\n",
    "        out_ids = [c.out_node_id for c in self.connection_genes if (c.in_node_id == node.id) and c.is_enabled]\n",
    "        return [n for n in nodes if n.id in out_ids]\n",
    "\n",
    "    def order_units(self, units):\n",
    "        \"\"\"\n",
    "        Implements a directed graph topological sort algorithm\n",
    "        Requires an acyclic graph - see _is_valid_connection method\n",
    "        :return: A sorted stack of NodeGene instances\n",
    "        \"\"\"\n",
    "\n",
    "        nodes = [u.ref_node for u in units]\n",
    "        visited = set()\n",
    "        ordered = []\n",
    "\n",
    "        for n in nodes:\n",
    "            if n not in visited:\n",
    "                self._order_units(n, nodes, ordered, visited)\n",
    "\n",
    "        ordered_units = []\n",
    "        for n in ordered:\n",
    "            for u in units:\n",
    "                if u.ref_node == n:\n",
    "                    ordered_units.append(u)\n",
    "                    break\n",
    "\n",
    "        return ordered_units\n",
    "\n",
    "    def _order_units(self, node, nodes, ordered, visited):\n",
    "        visited.add(node)\n",
    "\n",
    "        for out_node in self.get_outputs(node, nodes):\n",
    "            if out_node not in visited:\n",
    "                self._order_units(out_node, nodes, ordered, visited)\n",
    "\n",
    "        ordered.append(node)\n",
    "\n",
    "    def __str__(self):\n",
    "        ret = 'Connections:\\n\\n'\n",
    "        for connect_gene in self.connection_genes:\n",
    "            ret += str(connect_gene) + '\\n'\n",
    "\n",
    "        ret += 'Nodes: \\n\\n'\n",
    "        for node_gene in self.node_genes:\n",
    "            ret += str(node_gene) + '\\n'\n",
    "\n",
    "        return ret"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c998d138",
   "metadata": {},
   "source": [
    "## Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d7effdf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import copy\n",
    "\n",
    "import torch\n",
    "\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "def rand_uni_val():\n",
    "    \"\"\"\n",
    "    Gets a random value from a uniform distribution on the interval [0, 1]\n",
    "    :return: Float\n",
    "    \"\"\"\n",
    "    return float(torch.rand(1))\n",
    "\n",
    "\n",
    "def rand_bool():\n",
    "    \"\"\"\n",
    "    Returns a random boolean value\n",
    "    :return: Boolean\n",
    "    \"\"\"\n",
    "    return rand_uni_val() <= 0.5\n",
    "\n",
    "\n",
    "def get_best_genome(population):\n",
    "    \"\"\"\n",
    "    Gets best genome out of a population\n",
    "    :param population: List of Genome instances\n",
    "    :return: Genome instance\n",
    "    \"\"\"\n",
    "    population_copy = copy.deepcopy(population)\n",
    "    population_copy.sort(key=lambda g: g.fitness, reverse=True)\n",
    "\n",
    "    return population_copy[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e880dc0",
   "metadata": {},
   "source": [
    "## Crossover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c45620fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "This module contains crossover methods as described in Kenneth O. Stanley's NEAT\n",
    "paper.\n",
    "\n",
    "Todo:\n",
    "    * Allow other types of crossover?\n",
    "\n",
    "\"\"\"\n",
    "import logging\n",
    "from copy import deepcopy\n",
    "\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "def crossover(genome_1, genome_2, config):\n",
    "    \"\"\"\n",
    "    Crossovers two Genome instances as described in the original NEAT implementation\n",
    "    :param genome_1: First Genome Instance\n",
    "    :param genome_2: Second Genome Instance\n",
    "    :param config: Experiment's configuration class\n",
    "    :return: A child Genome Instance\n",
    "    \"\"\"\n",
    "\n",
    "    child = Genome()\n",
    "    best_parent, other_parent = order_parents(genome_1, genome_2)\n",
    "\n",
    "    # Crossover connections\n",
    "    # Randomly add matching genes from both parents\n",
    "    for c_gene in best_parent.connection_genes:\n",
    "        matching_gene = other_parent.get_connect_gene(c_gene.innov_num)\n",
    "\n",
    "        if matching_gene is not None:\n",
    "            # Randomly choose where to inherit gene from\n",
    "            if rand_bool():\n",
    "                child_gene = deepcopy(c_gene)\n",
    "            else:\n",
    "                child_gene = deepcopy(matching_gene)\n",
    "\n",
    "        # No matching gene - is disjoint or excess\n",
    "        # Inherit disjoint and excess genes from best parent\n",
    "        else:\n",
    "            child_gene = deepcopy(c_gene)\n",
    "\n",
    "        # Apply rate of disabled gene being re-enabled\n",
    "        if not child_gene.is_enabled:\n",
    "            is_reenabeled = rand_uni_val() <= config.CROSSOVER_REENABLE_CONNECTION_GENE_RATE\n",
    "            enabled_in_best_parent = best_parent.get_connect_gene(child_gene.innov_num).is_enabled\n",
    "\n",
    "            if is_reenabeled or enabled_in_best_parent:\n",
    "                child_gene.is_enabled = True\n",
    "\n",
    "        child.add_connection_copy(child_gene)\n",
    "\n",
    "    # Crossover Nodes\n",
    "    # Randomly add matching genes from both parents\n",
    "    for n_gene in best_parent.node_genes:\n",
    "        matching_gene = other_parent.get_node_gene(n_gene.id)\n",
    "\n",
    "        if matching_gene is not None:\n",
    "            # Randomly choose where to inherit gene from\n",
    "            if rand_bool():\n",
    "                child_gene = deepcopy(n_gene)\n",
    "            else:\n",
    "                child_gene = deepcopy(matching_gene)\n",
    "\n",
    "        # No matching gene - is disjoint or excess\n",
    "        # Inherit disjoint and excess genes from best parent\n",
    "        else:\n",
    "            child_gene = deepcopy(n_gene)\n",
    "\n",
    "        child.add_node_copy(child_gene)\n",
    "\n",
    "    return child\n",
    "\n",
    "\n",
    "def order_parents(parent_1, parent_2):\n",
    "    \"\"\"\n",
    "    Orders parents with respect to fitness\n",
    "    :param parent_1: First Parent Genome\n",
    "    :param parent_2: Secont Parent Genome\n",
    "    :return: Two Genome Instances\n",
    "    \"\"\"\n",
    "\n",
    "    # genotype.cpp line 2043\n",
    "    # Figure out which genotype is better\n",
    "    # The worse genotype should not be allowed to add excess or disjoint genes\n",
    "    # If they are the same, use the smaller one's disjoint and excess genes\n",
    "\n",
    "    best_parent = parent_1\n",
    "    other_parent = parent_2\n",
    "\n",
    "    len_parent_1 = len(parent_1.connection_genes)\n",
    "    len_parent_2 = len(parent_2.connection_genes)\n",
    "\n",
    "    if parent_1.fitness == parent_2.fitness:\n",
    "        if len_parent_1 == len_parent_2:\n",
    "            # Fitness and Length equal - randomly choose best parent\n",
    "            if rand_bool():\n",
    "                best_parent = parent_2\n",
    "                other_parent = parent_1\n",
    "        # Choose minimal parent\n",
    "        elif len_parent_2 < len_parent_1:\n",
    "            best_parent = parent_2\n",
    "            other_parent = parent_1\n",
    "\n",
    "    elif parent_2.fitness > parent_1.fitness:\n",
    "        best_parent = parent_2\n",
    "        other_parent = parent_1\n",
    "\n",
    "    return best_parent, other_parent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8843af6a",
   "metadata": {},
   "source": [
    "## Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "53c03d79",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import random\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "def mutate(genome, config):\n",
    "    \"\"\"\n",
    "    Applies connection and structural mutations at proper rate.\n",
    "    Connection Mutations: Uniform Weight Perturbation or Replace Weight Value with Random Value\n",
    "    Structural Mutations: Add Connection and Add Node\n",
    "    :param genome: Genome to be mutated\n",
    "    :param config: Experiments' configuration file\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "\n",
    "    if utils.rand_uni_val() < config.CONNECTION_MUTATION_RATE:\n",
    "        for c_gene in genome.connection_genes:\n",
    "            if utils.rand_uni_val() < config.CONNECTION_PERTURBATION_RATE:\n",
    "                perturb = utils.rand_uni_val() * random.choice([1, -1])\n",
    "                c_gene.weight += perturb\n",
    "            else:\n",
    "                c_gene.set_rand_weight()\n",
    "\n",
    "    if utils.rand_uni_val() < config.ADD_NODE_MUTATION_RATE:\n",
    "        genome.add_node_mutation()\n",
    "\n",
    "    if utils.rand_uni_val() < config.ADD_CONNECTION_MUTATION_RATE:\n",
    "        genome.add_connection_mutation()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdf1a6c5",
   "metadata": {},
   "source": [
    "## Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "eadd4899",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import sys\n",
    "\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "class Species:\n",
    "\n",
    "    def __init__(self, id, model_genome, generation):\n",
    "        self.id = id\n",
    "        self.model_genome = model_genome\n",
    "        self.members = []\n",
    "        self.fitness_history = []\n",
    "        self.fitness = None\n",
    "        self.adjusted_fitness = None\n",
    "        self.last_improved = generation\n",
    "\n",
    "    @staticmethod\n",
    "    def species_distance(genome_1, genome_2):\n",
    "        C1 = 1.0\n",
    "        C2 = 1.0\n",
    "        C3 = 0.5\n",
    "        N = 1\n",
    "\n",
    "        num_excess = genome_1.get_num_excess_genes(genome_2)\n",
    "        num_disjoint = genome_1.get_num_disjoint_genes(genome_2)\n",
    "        avg_weight_difference = genome_1.get_avg_weight_difference(genome_2)\n",
    "\n",
    "        distance = (C1 * num_excess) / N\n",
    "        distance += (C2 * num_disjoint) / N\n",
    "        distance += C3 * avg_weight_difference\n",
    "\n",
    "        return distance\n",
    "\n",
    "    @staticmethod\n",
    "    def stagnation(species, generation):\n",
    "        \"\"\"\n",
    "        From https://github.com/CodeReclaimers/neat-python/neat/stagnation.py\n",
    "        :param species: List of Species\n",
    "        :param generation: generation number\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        species_data = []\n",
    "        for s in species:\n",
    "            if len(s.fitness_history) > 0:\n",
    "                prev_fitness = max(s.fitness_history)\n",
    "            else:\n",
    "                # super small number\n",
    "                prev_fitness = -sys.float_info.max\n",
    "\n",
    "            s.fitness = max([g.fitness for g in s.members])\n",
    "            s.fitness_history.append(s.fitness)\n",
    "            s.adjusted_fitness = None\n",
    "\n",
    "            if prev_fitness is None or s.fitness >prev_fitness:\n",
    "                s.last_improved = generation\n",
    "\n",
    "            species_data.append(s)\n",
    "\n",
    "        # sort in ascending fitness order\n",
    "        species_data.sort(key=lambda g: g.fitness)\n",
    "\n",
    "        result = []\n",
    "        species_fitnesses = []\n",
    "        num_non_stagnant = len(species_data)\n",
    "        for i, s in enumerate(species_data):\n",
    "            # Override stagnant state if marking this species as stagnant would\n",
    "            # result in the total number of species dropping below the limit.\n",
    "            # Because species are in ascending fitness order, less fit species\n",
    "            # will be marked as stagnant first.\n",
    "            stagnant_time = generation - s.last_improved\n",
    "            is_stagnant = False\n",
    "            if num_non_stagnant > 1:\n",
    "                is_stagnant = stagnant_time >= 10\n",
    "\n",
    "            if (len(species_data) - i) <= 1:\n",
    "                is_stagnant = False\n",
    "\n",
    "            if (len(species_data) - i) <= 1:\n",
    "                is_stagnant = False\n",
    "\n",
    "            if is_stagnant:\n",
    "                num_non_stagnant -= 1\n",
    "\n",
    "            result.append((s, is_stagnant))\n",
    "            species_fitnesses.append(s.fitness)\n",
    "\n",
    "        return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61f38c97",
   "metadata": {},
   "source": [
    "## Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "0b19626c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "class Population:\n",
    "    __global_innovation_number = 0\n",
    "    current_gen_innovation = []  # Can be reset after each generation according to paper\n",
    "\n",
    "    def __init__(self, config):\n",
    "        self.Config = config()\n",
    "        self.population = self.set_initial_population()\n",
    "        self.species = []\n",
    "\n",
    "        for genome in self.population:\n",
    "            self.speciate(genome, 0)\n",
    "\n",
    "    def run(self):\n",
    "        for generation in range(1, self.Config.NUMBER_OF_GENERATIONS):\n",
    "            # Get Fitness of Every Genome\n",
    "            for genome in self.population:\n",
    "                genome.fitness = max(0, self.Config.fitness_fn(genome))\n",
    "\n",
    "            best_genome = get_best_genome(self.population)\n",
    "\n",
    "            # Reproduce\n",
    "            all_fitnesses = []\n",
    "            remaining_species = []\n",
    "\n",
    "            for species, is_stagnant in Species.stagnation(self.species, generation):\n",
    "                if is_stagnant:\n",
    "                    self.species.remove(species)\n",
    "                else:\n",
    "                    all_fitnesses.extend(g.fitness for g in species.members)\n",
    "                    remaining_species.append(species)\n",
    "\n",
    "            min_fitness = min(all_fitnesses)\n",
    "            max_fitness = max(all_fitnesses)\n",
    "\n",
    "            fit_range = max(1.0, (max_fitness-min_fitness))\n",
    "            for species in remaining_species:\n",
    "                # Set adjusted fitness\n",
    "                avg_species_fitness = np.mean([g.fitness for g in species.members])\n",
    "                species.adjusted_fitness = (avg_species_fitness - min_fitness) / fit_range\n",
    "\n",
    "            adj_fitnesses = [s.adjusted_fitness for s in remaining_species]\n",
    "            adj_fitness_sum = sum(adj_fitnesses)\n",
    "\n",
    "            # Get the number of offspring for each species\n",
    "            new_population = []\n",
    "            for species in remaining_species:\n",
    "                if species.adjusted_fitness > 0:\n",
    "                    size = max(2, int((species.adjusted_fitness/adj_fitness_sum) * self.Config.POPULATION_SIZE))\n",
    "                else:\n",
    "                    size = 2\n",
    "\n",
    "                # sort current members in order of descending fitness\n",
    "                cur_members = species.members\n",
    "                cur_members.sort(key=lambda g: g.fitness, reverse=True)\n",
    "                species.members = []  # reset\n",
    "\n",
    "                # save top individual in species\n",
    "                new_population.append(cur_members[0])\n",
    "                size -= 1\n",
    "\n",
    "                # Only allow top x% to reproduce\n",
    "                purge_index = int(self.Config.PERCENTAGE_TO_SAVE * len(cur_members))\n",
    "                purge_index = max(2, purge_index)\n",
    "                cur_members = cur_members[:purge_index]\n",
    "\n",
    "                for i in range(size):\n",
    "                    parent_1 = random.choice(cur_members)\n",
    "                    parent_2 = random.choice(cur_members)\n",
    "\n",
    "                    child = crossover(parent_1, parent_2, self.Config)\n",
    "                    mutate(child, self.Config)\n",
    "                    new_population.append(child)\n",
    "\n",
    "            # Set new population\n",
    "            self.population = new_population\n",
    "            Population.current_gen_innovation = []\n",
    "\n",
    "            # Speciate\n",
    "            for genome in self.population:\n",
    "                self.speciate(genome, generation)\n",
    "\n",
    "            if best_genome.fitness >= self.Config.FITNESS_THRESHOLD:\n",
    "                return best_genome, generation\n",
    "\n",
    "            # Generation Stats\n",
    "            if self.Config.VERBOSE:\n",
    "                logger.info(f'Finished Generation {generation}')\n",
    "                logger.info(f'Best Genome Fitness: {best_genome.fitness}')\n",
    "                logger.info(f'Best Genome Length {len(best_genome.connection_genes)}\\n')\n",
    "\n",
    "        return None, None\n",
    "\n",
    "    def speciate(self, genome, generation):\n",
    "        \"\"\"\n",
    "        Places Genome into proper species - index\n",
    "        :param genome: Genome be speciated\n",
    "        :param generation: Number of generation this speciation is occuring at\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "        for species in self.species:\n",
    "            if Species.species_distance(genome, species.model_genome) <= self.Config.SPECIATION_THRESHOLD:\n",
    "                genome.species = species.id\n",
    "                species.members.append(genome)\n",
    "                return\n",
    "\n",
    "        # Did not match any current species. Create a new one\n",
    "        new_species = Species(len(self.species), genome, generation)\n",
    "        genome.species = new_species.id\n",
    "        new_species.members.append(genome)\n",
    "        self.species.append(new_species)\n",
    "\n",
    "    def assign_new_model_genomes(self, species):\n",
    "        species_pop = self.get_genomes_in_species(species.id)\n",
    "        species.model_genome = random.choice(species_pop)\n",
    "\n",
    "    def get_genomes_in_species(self, species_id):\n",
    "        return [g for g in self.population if g.species == species_id]\n",
    "\n",
    "    def set_initial_population(self):\n",
    "        pop = []\n",
    "        for i in range(self.Config.POPULATION_SIZE):\n",
    "            new_genome = Genome()\n",
    "            inputs = []\n",
    "            outputs = []\n",
    "            bias = None\n",
    "\n",
    "            # Create nodes\n",
    "            for j in range(self.Config.NUM_INPUTS):\n",
    "                n = new_genome.add_node_gene('input')\n",
    "                inputs.append(n)\n",
    "\n",
    "            for j in range(self.Config.NUM_OUTPUTS):\n",
    "                n = new_genome.add_node_gene('output')\n",
    "                outputs.append(n)\n",
    "\n",
    "            if self.Config.USE_BIAS:\n",
    "                bias = new_genome.add_node_gene('bias')\n",
    "\n",
    "            # Create connections\n",
    "            for input in inputs:\n",
    "                for output in outputs:\n",
    "                    new_genome.add_connection_gene(input.id, output.id)\n",
    "\n",
    "            if bias is not None:\n",
    "                for output in outputs:\n",
    "                    new_genome.add_connection_gene(bias.id, output.id)\n",
    "\n",
    "            pop.append(new_genome)\n",
    "\n",
    "        return pop\n",
    "\n",
    "    @staticmethod\n",
    "    def get_new_innovation_num():\n",
    "        # Ensures that innovation numbers are being counted correctly\n",
    "        # This should be the only way to get a new innovation numbers\n",
    "        ret = Population.__global_innovation_number\n",
    "        Population.__global_innovation_number += 1\n",
    "        return ret"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a8abeed",
   "metadata": {},
   "source": [
    "## Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f25eb8c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from torch import autograd\n",
    "\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "\n",
    "class FeedForwardNet(nn.Module):\n",
    "\n",
    "    def __init__(self, genome, config):\n",
    "        super(FeedForwardNet, self).__init__()\n",
    "        self.genome = genome\n",
    "        self.units = self.build_units()\n",
    "        self.lin_modules = nn.ModuleList()\n",
    "        self.config = config\n",
    "        self.activation = a.Activations().get(config.ACTIVATION)\n",
    "\n",
    "        for unit in self.units:\n",
    "            self.lin_modules.append(unit.linear)\n",
    "\n",
    "    def forward(self, x):\n",
    "        outputs = dict()\n",
    "        input_units = [u for u in self.units if u.ref_node.type == 'input']\n",
    "        output_units = [u for u in self.units if u.ref_node.type == 'output']\n",
    "        bias_units = [u for u in self.units if u.ref_node.type == 'bias']\n",
    "        stacked_units = self.genome.order_units(self.units)\n",
    "\n",
    "        # Set input values\n",
    "        for u in input_units:\n",
    "            outputs[u.ref_node.id] = x[0][u.ref_node.id]\n",
    "\n",
    "        # Set bias value\n",
    "        for u in bias_units:\n",
    "            outputs[u.ref_node.id] = torch.ones((1, 1)).to(device)[0][0]\n",
    "\n",
    "        # Compute through directed topology\n",
    "        while len(stacked_units) > 0:\n",
    "            current_unit = stacked_units.pop()\n",
    "\n",
    "            if current_unit.ref_node.type != 'input' and current_unit.ref_node.type != 'bias':\n",
    "                # Build input vector to current node\n",
    "                inputs_ids = self.genome.get_inputs_ids(current_unit.ref_node.id)\n",
    "                in_vec = autograd.Variable(torch.zeros((1, len(inputs_ids)), device=device, requires_grad=True))\n",
    "\n",
    "                for i, input_id in enumerate(inputs_ids):\n",
    "                    in_vec[0][i] = outputs[input_id]\n",
    "\n",
    "                # Compute output of current node\n",
    "                linear_module = self.lin_modules[self.units.index(current_unit)]\n",
    "                if linear_module is not None:  # TODO: Can this be avoided?\n",
    "                    scaled = self.config.SCALE_ACTIVATION * linear_module(in_vec)\n",
    "                    out = self.activation(scaled)\n",
    "                else:\n",
    "                    out = torch.zeros((1, 1))\n",
    "\n",
    "                # Add to outputs dictionary\n",
    "                outputs[current_unit.ref_node.id] = out\n",
    "\n",
    "        # Build output vector\n",
    "        output = autograd.Variable(torch.zeros((1, len(output_units)), device=device, requires_grad=True))\n",
    "        for i, u in enumerate(output_units):\n",
    "            output[0][i] = outputs[u.ref_node.id]\n",
    "        return output\n",
    "\n",
    "    def build_units(self):\n",
    "        units = []\n",
    "\n",
    "        for n in self.genome.node_genes:\n",
    "            in_genes = self.genome.get_connections_in(n.id)\n",
    "            num_in = len(in_genes)\n",
    "            weights = [g.weight for g in in_genes]\n",
    "\n",
    "            new_unit = Unit(n, num_in)\n",
    "            new_unit.set_weights(weights)\n",
    "\n",
    "            units.append(new_unit)\n",
    "        return units\n",
    "\n",
    "\n",
    "class Unit:\n",
    "\n",
    "    def __init__(self, ref_node, num_in_features):\n",
    "        self.ref_node = ref_node\n",
    "        self.linear = self.build_linear(num_in_features)\n",
    "\n",
    "    def set_weights(self, weights):\n",
    "        if self.ref_node.type != 'input' and self.ref_node.type != 'bias':\n",
    "            weights = torch.cat(weights).unsqueeze(0)\n",
    "            for p in self.linear.parameters():\n",
    "                p.data = weights\n",
    "\n",
    "    def build_linear(self, num_in_features):\n",
    "        if self.ref_node.type == 'input' or self.ref_node.type == 'bias':\n",
    "            return None\n",
    "        return nn.Linear(num_in_features, 1, False)\n",
    "\n",
    "    def __str__(self):\n",
    "        return 'Reference Node: ' + str(self.ref_node) + '\\n'\n",
    "\n",
    "\n",
    "# TODO: Multiple GPU support get from config\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "896a4761",
   "metadata": {},
   "source": [
    "## Experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "330bd068",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Config:\n",
    "    DEVICE = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "    VERBOSE = True\n",
    "\n",
    "    NUM_INPUTS = 4\n",
    "    NUM_OUTPUTS = 4\n",
    "    USE_BIAS = True\n",
    "\n",
    "    ACTIVATION = 'sigmoid'\n",
    "    SCALE_ACTIVATION = 4.9\n",
    "\n",
    "    FITNESS_THRESHOLD = 100000.0\n",
    "\n",
    "    POPULATION_SIZE = 150\n",
    "    NUMBER_OF_GENERATIONS = 150\n",
    "    SPECIATION_THRESHOLD = 3.0\n",
    "\n",
    "    CONNECTION_MUTATION_RATE = 0.80\n",
    "    CONNECTION_PERTURBATION_RATE = 0.90\n",
    "    ADD_NODE_MUTATION_RATE = 0.03\n",
    "    ADD_CONNECTION_MUTATION_RATE = 0.5\n",
    "\n",
    "    CROSSOVER_REENABLE_CONNECTION_GENE_RATE = 0.25\n",
    "\n",
    "    # Top percentage of species to be saved before mating\n",
    "    PERCENTAGE_TO_SAVE = 0.80\n",
    "\n",
    "    def fitness_fn(self, genome):\n",
    "        # OpenAI Gym\n",
    "        env = gym.make('LongCartPole-v0')\n",
    "        done = False\n",
    "        observation = env.reset()\n",
    "\n",
    "        fitness = 0\n",
    "        phenotype = FeedForwardNet(genome, self)\n",
    "\n",
    "        while not done:\n",
    "            observation = np.array([observation])\n",
    "            input = torch.Tensor(observation).to(self.DEVICE)\n",
    "\n",
    "            pred = round(float(phenotype(input)))\n",
    "            observation, reward, done, info = env.step(pred)\n",
    "\n",
    "            fitness += reward\n",
    "        env.close()\n",
    "\n",
    "        return fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "d15d5a59",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'in_node_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[75], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtorch\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m neat \u001b[38;5;241m=\u001b[39m \u001b[43mPopulation\u001b[49m\u001b[43m(\u001b[49m\u001b[43mConfig\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m solution, generation \u001b[38;5;241m=\u001b[39m neat\u001b[38;5;241m.\u001b[39mrun()\n",
      "Cell \u001b[0;32mIn[72], line 19\u001b[0m, in \u001b[0;36mPopulation.__init__\u001b[0;34m(self, config)\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mspecies \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m genome \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpopulation:\n\u001b[0;32m---> 19\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspeciate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenome\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[72], line 109\u001b[0m, in \u001b[0;36mPopulation.speciate\u001b[0;34m(self, genome, generation)\u001b[0m\n\u001b[1;32m    102\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    103\u001b[0m \u001b[38;5;124;03mPlaces Genome into proper species - index\u001b[39;00m\n\u001b[1;32m    104\u001b[0m \u001b[38;5;124;03m:param genome: Genome be speciated\u001b[39;00m\n\u001b[1;32m    105\u001b[0m \u001b[38;5;124;03m:param generation: Number of generation this speciation is occuring at\u001b[39;00m\n\u001b[1;32m    106\u001b[0m \u001b[38;5;124;03m:return: None\u001b[39;00m\n\u001b[1;32m    107\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    108\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m species \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mspecies:\n\u001b[0;32m--> 109\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mSpecies\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspecies_distance\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenome\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mspecies\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel_genome\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mConfig\u001b[38;5;241m.\u001b[39mSPECIATION_THRESHOLD:\n\u001b[1;32m    110\u001b[0m         genome\u001b[38;5;241m.\u001b[39mspecies \u001b[38;5;241m=\u001b[39m species\u001b[38;5;241m.\u001b[39mid\n\u001b[1;32m    111\u001b[0m         species\u001b[38;5;241m.\u001b[39mmembers\u001b[38;5;241m.\u001b[39mappend(genome)\n",
      "Cell \u001b[0;32mIn[71], line 28\u001b[0m, in \u001b[0;36mSpecies.species_distance\u001b[0;34m(genome_1, genome_2)\u001b[0m\n\u001b[1;32m     26\u001b[0m num_excess \u001b[38;5;241m=\u001b[39m genome_1\u001b[38;5;241m.\u001b[39mget_num_excess_genes(genome_2)\n\u001b[1;32m     27\u001b[0m num_disjoint \u001b[38;5;241m=\u001b[39m genome_1\u001b[38;5;241m.\u001b[39mget_num_disjoint_genes(genome_2)\n\u001b[0;32m---> 28\u001b[0m avg_weight_difference \u001b[38;5;241m=\u001b[39m \u001b[43mgenome_1\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_avg_weight_difference\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenome_2\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     30\u001b[0m distance \u001b[38;5;241m=\u001b[39m (C1 \u001b[38;5;241m*\u001b[39m num_excess) \u001b[38;5;241m/\u001b[39m N\n\u001b[1;32m     31\u001b[0m distance \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m (C2 \u001b[38;5;241m*\u001b[39m num_disjoint) \u001b[38;5;241m/\u001b[39m N\n",
      "Cell \u001b[0;32mIn[67], line 167\u001b[0m, in \u001b[0;36mGenome.get_avg_weight_difference\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    165\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m c_gene \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnection_genes:\n\u001b[1;32m    166\u001b[0m     matching_gene \u001b[38;5;241m=\u001b[39m other\u001b[38;5;241m.\u001b[39mget_connect_gene(c_gene\u001b[38;5;241m.\u001b[39minnov_num)\n\u001b[0;32m--> 167\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mmatching_gene\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m!=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m:\n\u001b[1;32m    168\u001b[0m         weight_difference \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mfloat\u001b[39m(c_gene\u001b[38;5;241m.\u001b[39mweight) \u001b[38;5;241m-\u001b[39m \u001b[38;5;28mfloat\u001b[39m(matching_gene\u001b[38;5;241m.\u001b[39mweight)\n\u001b[1;32m    169\u001b[0m         num_weights \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "Cell \u001b[0;32mIn[67], line 54\u001b[0m, in \u001b[0;36mConnectionGene.__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__eq__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[0;32m---> 54\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39min_node_id \u001b[38;5;241m==\u001b[39m \u001b[43mother\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43min_node_id\u001b[49m) \u001b[38;5;129;01mand\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mout_node_id \u001b[38;5;241m==\u001b[39m other\u001b[38;5;241m.\u001b[39mout_node_id)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'in_node_id'"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "neat = Population(Config)\n",
    "solution, generation = neat.run()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
